### ISSUE TITLE
Improve logging in Yii 2.1

### ISSUE THREAD

**Issue by nkovacs** [yiisoft/yii2#13103]
The problems with the current `\yii\log\Target` and co.:

* `logVars` is not flexible enough. You can't access `Yii::$app->user->id`, for example.
* `getContextMessage` only supports returning a single string. This is inflexible for something like sentry, which allows you to pass structured data.
* The context message is logged as a separate message. This is not very useful for anything more complex than `FileTarget`. It's completely useless for [yii2-sentry](https://github.com/hellowearemito/yii2-sentry), so we had to disable `getContextMessage`.
* The first parameter to `Logger::log` can be an array, but that isn't really documented. You can't easily log extra data, you have to format it all into a string.

My proposal:

* `\yii\log\Target` allows you to attach extra data (an associative array) to each message. This can be configured as a callback or a method you can extend.
* `FileTarget` and `SyslogTarget` can use `VarDumper::dumpAsString` and append it to the message, so you get similar output as in Yii 2.0 (except it can be structured better, e.g. you could indent all the extra data so it's easier to read the log file).
* `DbTarget` can json encode it into a new field (or you can extend DbTarget and add new fields, allowing you to use db filtering, for example).
* `EmailTarget` can also use `VarDumper::dumpAsString`, but format it in a prettier way. Or you can extend it and do something fancier with the data.
* support for `logVars` and `getContextMessage`  can be kept for partial backwards compatibility, if `logVars` is configured, or `getContextMessage` is overridden in a child class, add their output to the extra data (I'm not sure if it's possible to be completely backwards compatible with Yii 2.0, probably not)
* `Logger::log` and `Yii::info` etc. allow you to pass extra data, which will be merged with the global extra data.

**Comment by samdark** [comment]
Yes, there are limitations which could be addressed in 2.1 branch. Do you want to work on it?

**Comment by nkovacs** [comment]
Yeah, I'll try to find some time to implement this.

**Comment by dynasource** [comment]
this issue should be taken into account https://github.com/yiisoft/yii2/issues/6472#issuecomment-288159529

**Comment by samdark** [comment]
#13702
